### PR TITLE
Add option to explicitly preprocess a batch before training

### PR DESCRIPTION
**Pull request by Rikorose**
## What does this PR do?
This PR adds a feature that enables explicit preprocessing of a batch. This is useful if the same preprocessing is necessary for `training_step()` and `validation_step()`. This will be further helpful for #293, when the preprocessing only needs to be done once per batch and for every tbptt step the preprocessed input can be reused.

Code example:
```py
def preprocess_batch(self, batch):
    batch.input_seq, lengths = nn.utils.rnn.pad_packed_sequence(
         batch.input_seq, batch_first=True
    )
    return batch

```

## PR review    
Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.   

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by williamFalcon** [comment]
what’s the difference between this and just processing the batch at the beginning of training_step and validation_step?

**Comment by Rikorose** [comment]
I am thinking of an implementation of #293. Here, `preprocess_batch` needs to be called once per batch. However, we do need to iterate over one the input and call `backward()` and `optimizer.step()` multiple times per batch. Thus, `preprocess_batch` cannot be called in the beginning of training_step since it would be called multiple times.

I see several options to resolve #293.
1. Use this PR and build an additional loop around the optimizer_closure + optimizer step ([L1334](https://github.com/williamFalcon/pytorch-lightning/blob/ac6d015/pytorch_lightning/trainer/trainer.py#L1334)-[L1399](https://github.com/williamFalcon/pytorch-lightning/blob/ac6d015/pytorch_lightning/trainer/trainer.py#L1399)), where we can iterate over the input sequence. The logic how often to iterate could be implemented in the pl module by adding a method like `tbptt_should_continue()`.
2. Provide a full implementation in trainer similar to [ignite's TBPTT Trainer](https://github.com/pytorch/ignite/blob/master/ignite/contrib/engines/tbptt.py).

**Comment by williamFalcon** [comment]
it’s much simpler to add a flag:

Trainer(bptt_steps=10) 

Or something like that.

**Comment by Rikorose** [comment]
That's true. The input might be very different depending on the use case, though. Do we have `torch.Tensor` or `torch.nn.utils.rnn.PackedSequence` as input? Does the user handle the different lengths in on batch? How many different inputs are there? One (Autoencoder like)? Two (Seq-Seq)? What if more? Do we use the same tbptt step size for all sequences?


I am currently thinking about an implementation using this PR to provide flexibility for different input types and provide an option i.e. as overwritable method in LightningModule to modify the optimization_closure(). 
Using this, optimizer.step() would only be called after the complete processing of one batch which is similar to gradient accumulation for multiple batches. In this case, will the gradients still need to be stored?

However, I am open for other implementation possibilities, that are able to implement `Trainer(bptt_steps=10)`.

**Comment by williamFalcon** [comment]
why don’t we start by writing out the vanilla implementations here (in this issue) for every case. 

then we can generalize

**Comment by Rikorose** [comment]
Let's discuss this in #293. I added a small script in a comment that showcases some of the variations.

**Comment by williamFalcon** [comment]
@Rikorose closing this because the solution will be part of the broader PR for tbptt
